File: arrays/intersectionOfTwoArrays.cpp
// 1) Sort arr1[] and arr2[]. This step takes O(mLogm + nLogn) time.
//2) Use O(m + n) algorithms to find union and intersection of two sorted arrays.

//Overall time complexity of this method is O(mLogm + nLogn).

/**
Intersection of two arrays Submissions: 5669   Accuracy: 26.54%   Difficulty: Basic   Marks: 1
            
Given two arrays with distinct elements A and B, print intersection (or common elements) of the two arrays.  If no element is common in two arrays, then print Zero.

Input:

The first line of input contains an integer T denoting the number of test cases.
The first line of each test case is N and M,N is the size of array A and M is size of array B.
The second line of each test case contains N input A[i].
The third line of each test case contains M input B[i].

Output:

Print the intersecting elements sorted in ascending order .If no element is common in two array, then print "Zero" without quotes.

Constraints:

1 ≤ T ≤ 50
1 ≤ N, M ≤ 100
1 ≤ A[i], B[i] ≤ 1000

Example:

Input:2
5 3
89 24 75 11 23
89 2 4
6 5
1 2 3 4 5 6
3 4 5 6 7

Output:
89
3 4 5 6
*/


#include <iostream>
#include<algorithm>
using namespace std;

void printIntersectionOfArrays(int *a, int *b, int m, int n) {
    int i=0, j=0;
    bool flag = false;
    // sort both arrays
    sort(a, a+m);
    sort(b, b+n);
    // print till both have common length available
    while(i<m && j<n) {
        if(a[i]<b[j]) {
            //cout<<a[i]<<" ";
            i++;
        } else if (a[i]>b[j]) {
            //cout<<b[j]<<" ";
            j++;
        } else {
            cout<<a[i]<<" ";
            i++;
            j++;
            flag = true;
        }
    }
    if(!flag)
        cout<<"Zero";
    cout<<endl;
}
int main() {
	int t, m,n;
	int arr1[1001], arr2[1001];
	cin>>t;
	while(t--) {
	    cin>>m>>n;
	    for(int i=0; i<m; i++) {
	        cin>>arr1[i];
	    }
	    for(int i=0; i<n; i++) {
	        cin>>arr2[i];
	    }
	    printIntersectionOfArrays(arr1, arr2, m, n);
	}
	return 0;
}

File: arrays/java/intersectionOfTwoArrays.java
/*
 Intersection of Two Arrays
Given two arrays, write a function to compute their intersection.

Example 1:

Input: nums1 = [1,2,2,1], nums2 = [2,2]
Output: [2]
Example 2:

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
Output: [9,4]
Note:

Each element in the result must be unique.
The result can be in any order.
*/
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> s = new HashSet<Integer>();
        for(int i=0; i<nums1.length; i++) {
            s.add(nums1[i]);
        }
        int j=0;
        for(int i=0; i<nums2.length; i++) {
            if (s.contains(nums2[i])) {
                nums2[j] = nums2[i];
                s.remove(nums2[i]);
                j++;
            }                
        }
        int res[] = new int[j];
        
        for(int i=0; i<j; i++) {
            res[i] = nums2[i];
        }
        return res;
    }
}